src/exchange/providers/base.py

```diff
@@ -28,3 +28,14 @@ def complete(
     ) -> Tuple[Message, Usage]:
         """Generate the next message using the specified model"""
         pass
+
+
+class MissingProviderEnvVariableError(Exception):
+    def __init__(self, env_variable: str, provider: str, instructions: Optional[str] = None) -> None:
+        self.env_variable = env_variable
+        self.provider = provider
+        self.instructions = instructions
+        self.message = f"Missing environment variable: {env_variable} for provider {provider}"
+        if instructions:
+            self.message += f". {instructions}"
```

Comment:
maybe rename instructions to instructions_url?

Also, not sure style, but if ok with multi-line error we could \n this and then make the asserts multi-line

```suggestion
            self.message += f". See {instructions} for instructions."
```

Reply:
Thanks @codefromthecrypt for your feedback!

Yes, we can rename to instructions_url so that we will have consistent message pattern.  On the hand, I am thinking whether we shall have some flexibility for the instructions. For example, people can put instructions without url.   It is hard to know the future :). 

I'll change as you suggested, in case we want to have more flexibility, it is fine because the function signature won't change anyway.